src/hotspot/os/aix/os_aix.cpp

```diff
@@ -1108,6 +1125,282 @@ bool os::dll_address_to_library_name(address addr, char* buf,
+  FILE *f = fopen(pgmpath, "r");
+  // read the generic XCOFF header and analyze the substructures
+  // to find the burned in libpath
+  fread(buffer, 1, FILHSZ_64 + _AOUTHSZ_EXEC_64, f);
+
+  if (((struct filehdr*)buffer)->f_magic == U802TOCMAGIC ) {
+    // __XCOFF32__
+    struct xcoffhdr xcoff;
+    struct scnhdr scn;
+    struct ldhdr ldr;
+    memcpy((char*)&xcoff, buffer, FILHSZ + _AOUTHSZ_EXEC);
+    int ldroffset = FILHSZ + xcoff.filehdr.f_opthdr + (xcoff.aouthdr.o_snloader -1)*SCNHSZ;
+    fseek (f, ldroffset, SEEK_SET);
+    fread(buffer, 1, SCNHSZ, f);
+    memcpy((char*)&scn, buffer, SCNHSZ);
+    if (strcmp(scn.s_name, ".loader")) {
+      buffer[0] = 0;
+      libpath = buffer;
+      fclose(f);
+      return libpath;
+    }
+    fseek (f, scn.s_scnptr, SEEK_SET);
+    fread(buffer, 1, LDHDRSZ, f);
+    memcpy((char*)&ldr, buffer, LDHDRSZ);
+    fseek (f, scn.s_scnptr + ldr.l_impoff, SEEK_SET);
+    fread(buffer, 1, 4096, f);
+  }
+  else if (((struct filehdr*)buffer)->f_magic == U64_TOCMAGIC ) {
+    // __XCOFF64__
+    struct _S_(xcoffhdr) xcoff64;
+    struct _S_(scnhdr) scn64;
+    struct _S_(ldhdr) ldr64;
+    memcpy((char*)&xcoff64, buffer, FILHSZ_64 + _AOUTHSZ_EXEC_64);
+    int ldroffset = FILHSZ_64 + xcoff64.filehdr.f_opthdr + (xcoff64.aouthdr.o_snloader -1)*SCNHSZ_64;
+    fseek (f, ldroffset, SEEK_SET);
+    fread(buffer, 1, SCNHSZ_64, f);
+    memcpy((char*)&scn64, buffer, SCNHSZ_64);
+    if (strcmp(scn64.s_name, ".loader")) {
+      buffer[0] = 0;
+      libpath = buffer;
+      fclose(f);
+      return libpath;
+    }
+    fseek (f, scn64.s_scnptr, SEEK_SET);
+    fread(buffer, 1, LDHDRSZ_64, f);
+    memcpy((char*)&ldr64, buffer, LDHDRSZ_64);
+    fseek (f, scn64.s_scnptr + ldr64.l_impoff, SEEK_SET);
+    fread(buffer, 1, 4096, f);
+  }
+  else
+    buffer[0] = 0;
+
+  fclose(f);
+  libpath = buffer;
+
+  return libpath;
+
+}
+
+// Simulate the library search algorithm of dlopen() (in os::dll_load)
+static bool search_file_in_LIBPATH(const char* path, struct stat64x* stat) {
+  if (path == nullptr)
+    return false;
+
+  char* path2 = os::strdup (path);
+  // if exist, strip off trailing (shr_64.o) or similar
+  char* substr;
+  if (path2[strlen(path2) - 1] == ')' && (substr = strrchr(path2, '('))) {
+    *substr = 0;
+  }
+
+  bool ret = false;
+  // If FilePath contains a slash character, FilePath is used directly,
+  // and no directories are searched.
+  // But if FilePath does not start with / or . we have to prepend it with ./
+  if (strchr(path2, '/')) {
+    stringStream combined;
+    if (*path2 == '/' || *path2 == '.')
+      combined.print("%s", path2);
+    else
+      combined.print("./%s", path2);
+    ret = (0 == stat64x(combined.base(), stat));
+    os::free (path2);
+    return ret;
+  }
+
+  const char* env = getenv("LIBPATH");
+  if (env == nullptr) {
+    // no LIBPATH, try with LD_LIBRARY_PATH
+    env = getenv("LD_LIBRARY_PATH");
+  }
+
+  stringStream Libpath;
+  if (env == nullptr) {
```

Comment:
Proposal for shorter version not needing string assembly:
```
const char* paths [2] = { env, rtv_linkedin_libpath() }:
for (int i = 0; i < 2; i ++) {
  const char* this_libpath = paths[i];
  if (this_libpath == nullptr) {
    continue;
  }
    ... do the token thing...
  }
}
```

Reply:
Sorry, I did not clearly understand how this should work. The mystery must be in _... do the token thing ..._